File: app/src/main/java/ru/wsr/stemobile/ui/main/SubstitutionKeyProvider.java
package ru.wsr.stemobile.ui.main;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.selection.ItemKeyProvider;

import ru.wsr.stemobile.data.model.Substitution;

public class SubstitutionKeyProvider extends ItemKeyProvider<Long> {

    private final SubstitutionAdapter mAdapter;

    SubstitutionKeyProvider(int scope, SubstitutionAdapter adapter) {
        super(scope);
        mAdapter = adapter;
    }

    @Nullable
    @Override
    public Long getKey(int position) {
        return (long) ((Substitution) mAdapter.getElements().get(position)).getID();
    }

    @Override
    public int getPosition(@NonNull Long key) {
        return mAdapter.getPositionByKey(key);
    }
}

File: app/src/main/java/ru/mrlargha/stemobile/ui/main/SubstitutionKeyProvider.kt
package ru.mrlargha.stemobile.ui.main

import androidx.recyclerview.selection.ItemKeyProvider
import ru.mrlargha.stemobile.data.model.Substitution

class SubstitutionKeyProvider internal constructor(
        scope: Int,
        private val mAdapter: SubstitutionAdapter
) : ItemKeyProvider<Long?>(scope) {
    override fun getKey(position: Int): Long? {
        return (mAdapter.elements!![position] as Substitution).iD.toLong()
    }

    override fun getPosition(key: Long): Int {
        return mAdapter.getPositionByKey(key)
    }

}